Change some wording in the solveset FAQ to clarify solve is not being…

klmcwhirter opened the pull request:
… removed

* full docs review is pending
* pushing this now to get the PR created and linked to issue 27332

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #27332 

#### Brief description of what is fixed or changed
This is a documentation change only. No code changes should be expected.

Change some wording suggesting that `solve` will be replaced by `solveset`.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
  * clarify wording in the docs that there are no current plans to remove `solve`
<!-- END RELEASE NOTES -->

Commit:
Change some wording in the solveset FAQ to clarify solve is not being removed

* full docs review is pending
* pushing this now to get the PR created and linked to issue 27332

Commit:
Adjust .mailcap entry and docs change formatting

klmcwhirter commented:
Note I am still reviewing the docs/doc strings/wiki page that need changing.

I just wanted to get incremental review started.

**I will post a comment when I believe I am done.**

Please do provide feedback, though, about whether release notes entry is needed or not. There is not a lot of value in it. But I am not familiar with the project policies yet.

Thanks.

Commit:
Reformat rst docs some more to address autodoc errors

klmcwhirter commented:
I still have a little more work to complete for the docs audit I am performing.

PLEASE DO NOT MERGE YET.

But please do provide feedback on:
* the need for release notes or not
* the wording of the change at this point

Thank you.

oscarbenjamin commented:
I think that the whole "What's wrong with solve():" section should be removed or completely rewritten.

The basic difference is that solve returns a list of substitutions which in practice is often what people want when calling a function like `solve`. The requirement to return a (finite) list of substitutions without the introduction of new symbols means that some cases are not well defined like `solve(sin(x), x)`. The `solveset` function has the capability to represent infinite solutions and other things but it does so at the expense of providing a list of substitutions meaning that its output is often not really useful.

Another difference is that solve is for multivariate systems and solveset has no capability for multivariate systems. Instead along with solveset there are linsolve and nonlinsolve but neither of those offers any particular advantage over solve (e.g. they don't actually handle infinite solution sets any better).

klmcwhirter commented:
Let's start with the section title: Differences Between solve and solveset.
Concise and less negative.

I'm not familiar enough with sympy to put together a set of feature lists.
But I would be happy to work any suggested prose through the process.
I guess I could start by simply rewriting what is there to be a less negative, more objective comparison and we could go from there. BTW, I expect this to be very iterative because I have not been using sympy that long.
But tomorrow :)
-k 
 
  On Sun, Dec 1, 2024 at 3:51 AM, Oscar ***@***.***> wrote:   


I think that the whole "What's wrong with solve():" section should be removed or completely rewritten.

The basic difference is that solve returns a list of substitutions which in practice is often what people want when calling a function like solve. The requirement to return a (finite) list of substitutions without the introduction of new symbols means that some cases are not well defined like solve(sin(x), x). The solveset function has the capability to represent infinite solutions and other things but it does so at the expense of providing a list of substitutions meaning that its output is often not really useful.

Another difference is that solve is for multivariate systems and solveset has no capability for multivariate systems. Instead along with solveset there are linsolve and nonlinsolve but neither of those offers any particular advantage over solve (e.g. they don't actually handle infinite solution sets any better).

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

smichr commented on doc/src/modules/solvers/solveset.rst:
```diff
@@ -476,10 +476,13 @@ What is the plan for solve and solveset?
 as solving nonlinear multivariate & LambertW type equations.  Hence, it's not
-yet a perfect replacement for ``solve``. As the algorithms in ``solveset``
-mature, ``solveset`` may be able to be used within ``solve`` to replace some of
+a perfect replacement for ``solve``. As the algorithms in ``solveset`` mature,
+``solveset`` may be able to be used within ``solve`` to replace some of
 its algorithms.
```
```suggestion
a perfect replacement for ``solve``.
```

smichr commented on doc/src/modules/solvers/solveset.rst:
```diff
@@ -476,10 +476,13 @@ What is the plan for solve and solveset?
 as solving nonlinear multivariate & LambertW type equations.  Hence, it's not
-yet a perfect replacement for ``solve``. As the algorithms in ``solveset``
-mature, ``solveset`` may be able to be used within ``solve`` to replace some of
+a perfect replacement for ``solve``. As the algorithms in ``solveset`` mature,
+``solveset`` may be able to be used within ``solve`` to replace some of
 its algorithms.
```
```suggestion
a perfect replacement for ``solve``.
```

smichr replied:
I would suggest just removing that 2nd sentence. How it is used where in SymPy is of no concern to the user.

klmcwhirter commented:
@oscarbenjamin ,
 > I think that the whole "What's wrong with solve():" section should be removed or completely rewritten.
> 

I lost track of the fact this is an FAQ. So ignore my section title change suggestion. I am just not a fan of finding fault with past design choices as a tool used to justify a companion piece of functionality.

After all, doesn't a carpenter use several types of hammers? You wouldn't expect a framing hammer to be used to do finishing work - would you? And so a carpenter does not have to justify usage of the finishing hammer by explaining "What is wrong with framing hammers?".

I do constantly get tripped up on the fact that solveset does not deal with multivariate equations.

Perhaps we should create another story to rewrite this whole document I think with better structure it might be much clearer and half the length.

Also, how do the guide pages fit into the strategy of documenting `solve`/`solveset` ? Much of the prose needed here seems to be duplicated there. I think these pages are intending to share the internal design (in user friendly language) whereas the guide pages are intended to share "how to use" `solve` / `solveset`. Is that right?

My point is this starts to feel like severe scope creep to include as part of this story. The PM in me is wanting to create another story for restructuring the docs.

This simple story has served the purpose of getting me familiar with how to contribute to this project (think .mailcap / authors - a bit of surprising complexity I have not encountered before). I now know how to build the docs using docker. Although that is not successful for me without making changes to a few files to workaround warnings being treated by `make html` as errors.

And while I do have quite a bit of technical writing experience; docs rewrite really was not the scope of this story.

I'm in the process of addressing the feedback from @smichr. I'll try to make some minor changes along the lines of your suggestions, but will keep myself focused on the goal of "clarifying that solveset is not replacing solve". 

OK?

> The basic difference is that solve returns a list of substitutions which in practice is often what people want when calling a function like `solve`. The requirement to return a (finite) list of substitutions without the introduction of new symbols means that some cases are not well defined like `solve(sin(x), x)`. The `solveset` function has the capability to represent infinite solutions and other things but it does so at the expense of providing a list of substitutions meaning that its output is often not really useful.
> 
> Another difference is that solve is for multivariate systems and solveset has no capability for multivariate systems. Instead along with solveset there are linsolve and nonlinsolve but neither of those offers any particular advantage over solve (e.g. they don't actually handle infinite solution sets any better).

oscarbenjamin commented:
I think we mostly just need to remove things from the solveset documentation that claim that it is better than solve or that there is something wrong with solve. The guides already explain the differences reasonably well and so the solveset docs page can just explain solveset while also having a link to the guides.

klmcwhirter commented:
> I think we mostly just need to remove things from the solveset documentation that claim that it is better than solve or that there is something wrong with solve. The guides already explain the differences reasonably well and so the solveset docs page can just explain solveset while also having a link to the guides.

Perfect. I had that thought too. I am going to push up the change for @smichr first and then will work on your suggestions. Watch for those ...

But I do think the docs could use more work. That is a much larger task. I would be up for it, but after I finish the project that led me to `sympy` in the first place: [pi-day-2025-with-py](https://github.com/klmcwhirter/pi-day-2025-with-py) ;)

Commit:
Address review comments from @smichr

oscarbenjamin commented:
> I do think the docs could use more work.

I agree but we can make small improvements as well.

klmcwhirter commented:
@smichr I reworded the Plans section completely. See the [ci/circleci](https://output.circle-artifacts.com/output/job/7f4c9fd9-83ea-437d-b846-8c8e031fee90/artifacts/0/doc/_build/html/modules/solvers/solveset.html#what-is-the-plan-for-solve-and-solveset) generated docs.

This is in anticipation of other changes that I am working on with @oscarbenjamin. The answer to this question can be more concise.

Do you agree with my change here?

Commit:
draft of changes to the first solveset FAQ

klmcwhirter commented:
@oscarbenjamin 

I pushed a draft of putting in place some of your suggestions for the first FAQ. I do think a lot (or all) of what is in the first bullet point should be merged with the Explanation section in the [solve docstring](https://github.com/sympy/sympy/blob/master/sympy/solvers/solvers.py#L366).

Thoughts?

oscarbenjamin commented:
> I do think a lot (or all) of what is in the first bullet point should be merged with the Explanation section in the [solve docstring](https://github.com/sympy/sympy/blob/master/sympy/solvers/solvers.py#L366) or perhaps in the [Solve Output by Type](https://output.circle-artifacts.com/output/job/2d8903c1-762c-4710-ac2f-d2bec6eccfbd/artifacts/0/doc/_build/html/explanation/solve_output.html) page

Yes, it shouldn't be in the `solveset` page I think but same goes for the other bullet points. Also the top of the page should not say the "new module". I think this page can just say that `solveset` returns a `Set` representing the solutions of a univariate equation and then give the examples that it has for `solveset` below.

> Point 3 seems out of place to me. minima / maxima is a specialized use case. Not sure it belongs in the list here.

The point is expressed badly but what it is really saying is an important fundamental difference between `solve` and `solveset`:

There is not a clear way in the interface of `solve` to communicate whether or not all possible solutions to the system were found or that there are provably no solutions or that the real set of solutions should be finite or infinite etc. In some cases `solve` may return a few solutions when others (potentially infinitely many) also exist. It is to some extent inherent to the API of solve that it does not have a way to represent these different cases distinctly.

In other cases `solve` might return no solutions as an empty list but that might just mean that it couldn't find the solutions rather than being a definitive statement that no solutions exist. That is potentially just something that should be considered a bug in solve though since it does have a way to indicate that it can't solve the equations by raising an error:
```
In [2]: solve(2**x + 3**x - 3, x)
---------------------------------------------------------------------------
NotImplementedError
```

klmcwhirter commented:
> Yes, it shouldn't be in the `solveset` page I think but same goes for the other bullet points. Also the top of the page should not say the "new module". I think this page can just say that `solveset` returns a `Set` representing the solutions of a univariate equation and then give the examples that it has for `solveset` below.

Aaah - good point.

> There is not a clear way in the interface of `solve` to communicate whether or not all possible solutions to the system were found or that there are provably no solutions or that the real set of solutions should be finite or infinite etc. In some cases `solve` may return a few solutions when others (potentially infinitely many) also exist. It is to some extent inherent to the API of solve that it does not have a way to represent these different cases distinctly.
> ...
> NotImplementedError

So that is what is meant by the "We don't Know" statement. You are right, this is an important thing to highlight.

Thanks for the feedback.

I will work on this some more before I tackle thinking about complimentary enhancements to the [Solve Output by Type](https://output.circle-artifacts.com/output/job/2d8903c1-762c-4710-ac2f-d2bec6eccfbd/artifacts/0/doc/_build/html/explanation/solve_output.html) page.

I'll push another draft up later today.

Commit:
Reword top of solveset.rst some more based on review feedback

* feedback from @oscarbenjamin
* changes through Why solveset()? section
* added Univariate Equations section to solvers.rst to further
communicate that solveset is a first-class solver

klmcwhirter commented:
@oscarbenjamin, I pushed another draft.

Changes in:
* [solveset.rst](https://output.circle-artifacts.com/output/job/6af85be8-b0bd-450d-9df6-f6264340af41/artifacts/0/doc/_build/html/modules/solvers/solveset.html) through the Why solveset()? section

* added to [solvers.rst](https://output.circle-artifacts.com/output/job/6af85be8-b0bd-450d-9df6-f6264340af41/artifacts/0/doc/_build/html/modules/solvers/solvers.html#univariate-equations) - Univariate Equations section (for completeness) to further communicate that solveset and solve are peers - although I welcome word-smithing feedback on the section name. Univariate Equations is just my best guess based on the stated purpose of the solveset function and the other section names.

I pulled all the following stuff out from solveset.rst that needs a decision.

Should it be kept? I am not entirely sure it would be missed. But someone put some work into enumerating these cases and so I did not want to just remove it without asking.

If so, should I find a way to merge it into [Solve Output by Type](https://docs.sympy.org/latest/explanation/solve_output.html) as I suggested earlier? 

```
From solveset.rst - needs a place

   * Single solution : `x = 1`
   * Multiple solutions: `x^2 = 1`
   * No Solution: `x^2 + 1 = 0 ; x \in \mathbb{R}`
   * Interval of solution: `\lfloor x \rfloor = 0`
   * Finite list (even though there are infinitely many solutions): `\sin(x) = 0`

     *Compare the* ``solve`` *output of* `[0, \pi ]` *to the* ``solveset`` *output shown in the* :ref:`why-solveset` *section below.*

   * Multivariate functions with point solutions: `x^2 + y^2 = 0`
   * Multivariate functions with non-point solution: `x^2 + y^2 = 1`
   * System of equations: `x + y = 1` and `x - y = 0`
   * Relational: `x > 0`
   * And the most important case: "We don't Know"


3. There are cases like finding the maxima and minima of function using
   critical points where it is important to know if it has returned all the
   solutions. ``solve`` does not guarantee this.
```

I will check back tomorrow. Need to put this down for awhile.

But it does feel like we are making progress. 

Thanks again for the help with this.

oscarbenjamin commented on doc/src/modules/solvers/solveset.rst:
```diff
@@ -474,11 +496,12 @@ How do we deal with cases where only some of the solutions are known?
 
-There are still a few things ``solveset`` can't do, which ``solve`` can, such
-as solving nonlinear multivariate & LambertW type equations.  Hence, it's not
-yet a perfect replacement for ``solve``. As the algorithms in ``solveset``
-mature, ``solveset`` may be able to be used within ``solve`` to replace some of
-its algorithms.
+Since the goals for ``solveset`` are different from ``solve``,
+``solveset`` is not intended to be a perfect replacement
+for ``solve``. Rather, ``solveset`` should be viewed as complimentary.
+
+As such, there are no current plans to remove ``solve``.
```
I think we should just remove this section.

klmcwhirter replied:
That is truly ironic - this is the very section that contained the word 'yet' that started all of this.

But you are right. Once the other items we have been working were addressed, the existence of this section does not make sense.

Consider it on the cutting floor.

oscarbenjamin commented:
> I pulled all the following stuff out from solveset.rst that needs a decision.
> 
> Should it be kept? I am not entirely sure it would be missed.

I think let's just get rid of it.

klmcwhirter commented on doc/src/modules/solvers/solveset.rst:
```diff
@@ -6,62 +6,83 @@ Solveset
+
+* It has a generalized output to handle various types of solutions.
+  It needs to return a lot of types of solutions consistently. It can deal with
+  univariate as well as multivariate equations as well as various other sorts
```
This clumsy wording - various other sorts of variability (ick!) - will be revised in the next draft.

Commit:
Address review comments and some clumsy wording

klmcwhirter commented:
@oscarbenjamin , @smichr ,

OK - pushed what I think is a candidate to wrap this up.

I reworded some clumsy wording (my bad), and removed the 'What is the plan ...' section.

Here is the link to the [circleci generated page](https://output.circle-artifacts.com/output/job/bab9034f-8f88-4082-a731-54bef7031ef3/artifacts/0/doc/_build/html/modules/solvers/solveset.html) for this build.


If this passes your review, should I squash the commits? How do you feel about that?

I would need to recreate the branch and PR.

What do you think? Do you care that there are 7 incremental commits?

Please let me know. I am OK either way.

oscarbenjamin commented:
I think that this is an improvement as is and can be merged. I would squash the commits but it doesn't really matter.

There is still plenty more to do. Looking at it I think that maybe everything at the top of this page should just be moved somewhere else. At least with the changes here it doesn't imply that solve should not be used.

klmcwhirter commented on doc/src/modules/solvers/solveset.rst:
```diff
@@ -6,62 +6,84 @@ Solveset
 
-What's wrong with solve():
---------------------------
+How solveset() is different from solve()
+----------------------------------------
+
+SymPy already has a pretty powerful ``solve`` function. But it has different
+goals. For example:
+
+* It has a generalized output interface to handle various types of solutions.
+  It needs to return many types of solutions consistently. It can deal with
+  univariate as well as multivariate equations as well as other kinds
+  of variability due to the system of equations and options passed into it,
+  producing associated outputs. See :ref:`solve_output` for more details.
+
+* The input API has a lot of parameters. It is a facade around
+  more specialized solvers. *This feature of* ``solve`` *could be viewed as analogous
+  to how* ``simplify`` *is a facade around more specialized simplification functions.*
+
+  See :func:`~sympy.solvers.solvers.solve` for more details about the available parameters
+  and output options.
```
By `criticizing poor development choices with solve` are you referring to the use of the word `consistently`?

Because I do not hear critique in that section. Now, I certainly do not have the history with the solvers code base that you 2 do.

What I do feel important is calling out the multivariate difference. I have been bit by that multiple times in the last year because I just refused to accept that solveset was truly for univariate equations if it was intended to be a replacement for solve.

I do hear what you are saying, but ...
* it feels like most of that stuff is covered in the next sections
* I have been trying to keep this `differences` section on the light side, adding links to other pages focusing on solve as I went
* the last 2 bullet points were from you - I do realize that quoting you verbatim may not have been what you really wanted - let's refine it. What I do feel strongly about in this blurb is readability. If I do not use bullets the average user will gloss over the points being presented and not read them; missing the full effect. But, with bullets, they can always come back later to review as reference.

What I was hoping for in this first section was half a page of text; brief, concise, to the point. And then get on with the docs for solveset where those kinds of usage examples would appear.

The `Why solveset()?` section has the `sin(x) = 0` example, which is a good one. Is there anything missing that you would like to add?

Should we add the first example (to the same section) to demonstrate the partial solution feature you are highlighting?

With that in mind how can I further simplify the differences section so it is useful to a user?

It feels pretty good to me. It is relatively light and with the new links in place adds to the discoverability.

But please do poke holes in how I see things.

To be honest, I am more concerned about your observation about `criticizing ...` than I am the missing example.

oscarbenjamin replied:
The text is using various euphemisms to say that `solve` has a badly designed and inconsistent interface. there is no reason it needs to be so badly designed, just people making poor choices that should have been fixed a long time ago but now it's too difficult because of compatibility:

It should always return a list of dicts:
```python
In [7]: solve(x-1)
Out[7]: [1]

In [8]: solve([x-1])
Out[8]: {x: 1}

In [9]: solve(x**2-1)
Out[9]: [-1, 1]

In [10]: solve([x**2-1])
Out[10]: [{x: -1}, {x: 1}]
```
This isn't a "generalised interface", it is just bad design. It should have been obvious that there would be a need to return multiple solutions (hence a list) and that it can't just be a list of numbers if you want multivariate equations (hence a list of dicts).

The basic format cannot handle inequalities but someone thought it would be good to add that and have it return a completely different type:
```python
In [11]: solve(x**2 < 1)
Out[11]: -1 < x ∧ x < 1

In [12]: solve([x**2-1, x > 0])
Out[12]: x = 1
```
I think most people would want the latter case to return `[{x: 1}]` but if you have inequalities then it returns a Boolean instead of a list of solutions. It was a mistake to add this to `solve` especially since it prevents a more useful interpretation of what inequalities could be used for.

There is a weird handling of equations involving booleans that doesn't really work properly:
```python
In [14]: solve(Eq(Not(x), True), x)
Out[14]: False
```
This should just raise an error instead.

In some situations solve returns an "undetermined coefficients" solution:
```python
In [20]: solve(x**2*y + z, [y, z])
Out[20]: {y: 0, z: 0}

In [21]: solve([x**2*y + z], [y, z])
Out[21]:
⎧   -z ⎫
⎪y: ───⎪
⎨    2 ⎬
⎪   x  ⎪
⎩      ⎭
```
There is a separate function for undetermined coefficients which should be used instead. The nature of the computed solution should not depend on whether there is a single equation or a list of one equation.

None of these are good ideas: it should just always return a list of dicts and it should be for solving equations but perhaps with some support for inequalities that filter the equations.

However none of these things is a reason to use solveset. If you want a consistent return type then you can just always pass a list of equations and use dict=True. The reasons listed here are reasons why sympy developers at the time thought that interface design should be better but they are not reasons for wanting to use mathematical sets instead of a list of dicts.

I agree that multivariate handling should be mentioned since that is an important usability difference. Note that along with solveset are linsolve and nonlinsolve which are supposed to handle multivariate equations but have been misdesigned so that they don't improve on solve in any fundamental way (they don't actually use mathematical sets properly).

klmcwhirter commented:
> I think that this is an improvement as is and can be merged. I would squash the commits but it doesn't really matter.
> 
> There is still plenty more to do. Looking at it I think that maybe everything at the top of this page should just be moved somewhere else. At least with the changes here it doesn't imply that solve should not be used.

Let me work your partial solution example into place at least before I squash. I think that is super valuable. I honestly did not understand how much until I saw it.

Commit:
Add partial solution example to the Why solveset()? section

klmcwhirter commented:
@oscarbenjamin when you get time, take a look at the bullet I added to the end of the `Why solveset()?` section.

[circleci generated page](https://output.circle-artifacts.com/output/job/c86db308-ea3d-4812-a2da-4524d91e97eb/artifacts/0/doc/_build/html/modules/solvers/solveset.html#why-solveset) for this build.

I do remain concerned about your observation. Do you think that is just because of your history with the code and docs? Or is there some low hanging fruit that can be plucked to address your concern?

I'm already setup to build the docs. I will hang on to my custom Dockerfile though just in case - either way.

oscarbenjamin commented:
> I do remain concerned about your observation. Do you think that is just because of your history with the code and docs? Or is there some low hanging fruit that can be plucked to address your concern?

This all predates my own involvement with sympy but it is quite clear that this text was written to proselytise which is fine but doesn't belong in the user-facing documentation especially not many years later when it is obvious that solveset does not replace solve and that solve is not going to be removed.

The fact that that proselytising has been left there so long makes it particularly important that we don't leave anything that resembles it in the docs. Many people will remember these things for a long time.

klmcwhirter commented:
> The fact that that proselytising has been left there so long makes it particularly important that we don't leave anything that resembles it in the docs. Many people will remember these things for a long time.

So give me some hints about what you are seeing. I am probably a little too close to the material at this point. After reviewing so much of the docs I may have lost my objectivity to some degree at least.

oscarbenjamin commented:
I think that the whole

> How solveset() is different from solve()

section doesn't really need to be there if this is the page that explains `solveset`. Other guides can (and already do) explain when you might use one or the other and can be linked at the top.

There are things like

> SymPy already has a pretty powerful solve function.

What does "already" mean? Note that this was written something like 10 years ago probably when solveset was new.

> It has a generalized output interface to handle various types of solutions.

This is a euphemism for having inconsistent badly designed API.

> The input API has a lot of parameters.

It is true that solve has a lot of parameters but that doesn't need to be mentioned here.

> Why Solveset?

I don't think we need this question. We can just say what solveset is and someone can decide if that is what they want.

> solveset has a consistent input and output interface:

Things like this are not usually mentioned in the docs.

Most of the things that compare with solve could just be statements about solveset without needing any comparison to solve.

> solveset has simpler input API, unlike solve. 

> solveset removes the flags argument of solve, which had made the input API more complicated and output API inconsistent.

I don't even know why we have all of these headings and this big preamble. Most of this could just be single doctest examples in the solveset docstring with no mention of `solve`. The main thing this page should explain is what is solveset and how do you use it. Other pages like [simplify](https://docs.sympy.org/latest/modules/simplify/simplify.html) in the reference docs don't have an essay about why the functions were added and how much better they are than the previous versions.

klmcwhirter commented:
> I don't even know why we have all of these headings ...

See? This is what I was saying as we got started. That is why I wanted to just yank out the word `yet` and let it go. This is a major can of worms.

In order to learn how to `solve` with sympy, I need to refer to 4 different major sections: Tutorials, How-to Guides, Explanations and API Reference.

And then taking another step back, why after reading the Installation section am I dump;ed right into Tutorials?

Where is the Architectural or at least Concepts introduction?

So make the executive call. I am up for whatever you want to address.

BTW, when is the proposed release date for 1.14 so I can keep that in mind.

I think we can handle some scope creep; how much depends on the release schedule.

oscarbenjamin commented:
Firstly I think that this is fine to merge as is: we don't need to fix everything all at once.

If we are thinking about something bigger then I wonder if @asmeurer or @bertiewooster has a view here on the solveset page or your more general point about the structure of the docs.

klmcwhirter commented:
Let's just call it then. How did you like the bullet I added about the partial solutions?

Should I include that commit in what I squash?

oscarbenjamin commented:
> How did you like the bullet I added about the partial solutions?

It looks fine to me.

klmcwhirter commented:
OK I'll include it. It is a really good example. Thanks.

bertiewooster commented:
I haven't reviewed the docs recently, so take my comments for what they're worth.

> why after reading the Installation section am I dumped right into Tutorials?
> Where is the Architectural or at least Concepts introduction?

Most new users typically want to get started using a package as quickly as possible, so Installation being followed by Tutorials matches their desired path. Adding a link on the Tutorials page to an Architectural or at least Concepts introduction (if or when such pages exist) seems like a reasonable idea.

As far as the [solveset page](https://docs.sympy.org/latest/modules/solvers/solveset.html), it seems a fair point that it need not compare solveset to solve. Perhaps such comparisons could be moved to a page in the [Solve Equations](https://docs.sympy.org/latest/guides/solving/index.html) section, quite possibly as a section on the page [Solving Guidance](https://docs.sympy.org/latest/guides/solving/solving-guidance.html), and dispassionately recast as why you might choose one instead of the other based on your needs.
